File: TestMoshi.java
import com.squareup.moshi.Moshi;
import com.squareup.moshi.JsonAdapter;

public class TestMoshi {
    public static class TestBook {
        public String id;
        public String title;
    }
    public static void main(String[] args) throws Exception {
        Moshi moshi = new Moshi.Builder().build();
        JsonAdapter<TestBook> adapter = moshi.adapter(TestBook.class);
        TestBook b = adapter.fromJson("{\"id\": 123, \"title\": \"Hello\"}");
        System.out.println("Success: " + b.id + " " + b.title);
    }
}

File: app/src/main/java/com/example/TestMoshi.kt
package com.example

import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class TestBook(
    val id: String,
    val title: String? = null
)

fun testMoshiParser() {
    val moshi = Moshi.Builder().addLast(KotlinJsonAdapterFactory()).build()
    val adapter = moshi.adapter(TestBook::class.java)
    try {
        val b = adapter.fromJson("{\"id\": 123, \"title\": \"Hello\"}")
        println("Success: $b")
    } catch (e: Exception) {
        println("Error: ${e.message}")
    }
}
